### PR TITLE
Simplified unpack with additional error messages

### PR DESCRIPTION
**Pull request by Stonesjtu**
The original version doesn't work for `torch.Tensor`, it raises **Runtime Error: Overflow when unpacking long** from the C backend:
https://github.com/pytorch/pytorch/blob/6204877cd44594541956d164e324642cae523d59/torch/csrc/utils/python_numbers.h#L48-L58

here's a minimal snippet:

```python
import torch
a = torch.Tensor(5)
print(a)
```

Actually I don't quite understand the existing method, it's not numerical stable because some random float numbers may satisfy `value == math.ceil(value)` in python.

### CODE REVIEW

**Comment by apaszke** [comment]
I don't think this is a complete fix. I've seen the same error message show up in other circumstances, and I've finally found a good repro script:
```python
torch.Tensor(5).fill_(math.floor(1e+40))
```
This is basically the same problem. `math.floor` returns an int, and because of this it gets rerouted to `unpackLong`, while it is used on an operation with a double tensor, and it is a value representable in the double format. You can check that:
```python
float(math.floor(1e+40)) == 1e+40
```

**Comment by Stonesjtu** [comment]
Let's make the inner process clear:
```python
v = 1e+40 # float
a = int(1e+40) # a large number int in python, but not representable by long (10^40 > 2^63)
unpacklong(a) # overflow happens
```

I think adding a type conversion to the item type of target tensor should fix the issue:
```python
def fill_(value):
    target_value = target_type(value)
    # processing with target_value
```

**Comment by ezyang** [comment]
This looks fine to me. Can we get a test for the thing that was originally failing?

**Comment by Stonesjtu** [comment]
After several tests, the `torch.Tensor([1])` constructor dispatches to the `unpackdouble`, but `torch.Tensor(4).fill_(1)` dispatches to the `unpacklong`. Could someone explain it for me?

File: torch/csrc/utils/python_numbers.h [hunk 1]
```diff
@@ -66,28 +77,17 @@ inline bool THPUtils_checkDouble(PyObject* obj) {
...
+    throw python_error();
   }
+  // raise warning when converting a long to float with precision loss
   if (PyLong_Check(obj)) {
-    int overflow;
-    long long value = PyLong_AsLongLongAndOverflow(obj, &overflow);
-    if (overflow != 0) {
-      throw std::runtime_error("Overflow when unpacking double");
-    }
     if (value > DOUBLE_INT_MAX || value < -DOUBLE_INT_MAX) {
```
Comment (apaszke) [inline torch/csrc/utils/python_numbers.h]: How is `value` supposed to be out of range of `double` if it's a double itself?

Comment (Stonesjtu) [reply torch/csrc/utils/python_numbers.h]: The original overflow message is not needed here.

early version:
`unpackDouble(python_int) --> c_long_long(python_int) --> c_double(c_long_long)`
unpackDouble(int(1e+40)) would raise the **overflow** error because long long  (typically 64 bits) can only represent up to **2^127**.  However the 1e+40 is totally representable by a `double` type.

Using double to represent `int(1e+40)` causes problem of **precision loss** as the absolute value is too large (here `DOUBLE_INT_MAX`). Compared with overflow, I think **precision loss** is somehow acceptable so a *warning* will be raised instead of *error*

this version:
`unpackDouble(python_int) --> python_float(python_int) --> c_double(python_float)`
The type conversions are all handled by cpython, so we can simply check `PyErr` and return `NULL` on error.

**Comment by Stonesjtu** [comment]
I think the failed test is not caused by this PR. Need reviews plz.

**Comment by ezyang** [comment]
> After several tests, the torch.Tensor([1]) constructor dispatches to the unpackdouble, but torch.Tensor(4).fill_(1) dispatches to the unpacklong. Could someone explain it for me?

That's pretty weird. Since you made that comment, we've merged Variable and Tensor, so maybe this is more uniform now. Can you try it again? @gchanan do you know off the top of your head what the behavior here should be?

**Comment by zou3519** [comment]
I don't think this is a problem anymore:
```
In [1]: import torch
   ...: a = torch.Tensor(5)
   ...: print(a)

tensor([1.3401e-04, 3.0735e-41, 4.9656e+28, 4.5439e+30, 5.7453e-44])
```
Please feel free to reopen this if it is.
